federation/pkg/kubefed/init/init.go

```diff
@@ -0,0 +1,502 @@
+	// Build the secret object with API server credentials.
+	secret := &api.Secret{
+		ObjectMeta: api.ObjectMeta{
+			Name:      credentialsName,
+			Namespace: namespace,
+		},
+		Data: map[string][]byte{
+			"ca.crt":     certutil.EncodeCertPEM(entKeyPairs.ca.Cert),
+			"server.crt": certutil.EncodeCertPEM(entKeyPairs.server.Cert),
+			"server.key": certutil.EncodePrivateKeyPEM(entKeyPairs.server.Key),
+		},
+	}
+
+	// Boilerplate to create the secret in the host cluster.
+	return clientset.Core().Secrets(namespace).Create(secret)
+
+}
+
+func createControllerManagerKubeconfigSecret(clientset *client.Clientset, namespace, name, svcName, kubeconfigName string, entKeyPairs *entityKeyPairs) (*api.Secret, error) {
+	basicClientConfig := kubeadmutil.CreateBasicClientConfig(
+		name,
+		fmt.Sprintf("https://%s", svcName),
+		certutil.EncodeCertPEM(entKeyPairs.ca.Cert),
+	)
+
+	config := kubeadmutil.MakeClientConfigWithCerts(
+		basicClientConfig,
+		name,
+		"federation-controller-manager",
+		certutil.EncodePrivateKeyPEM(entKeyPairs.cm.Key),
+		certutil.EncodeCertPEM(entKeyPairs.cm.Cert),
+	)
+
+	return util.CreateKubeconfigSecret(clientset, config, namespace, kubeconfigName, false)
+}
+
+func createPVC(clientset *client.Clientset, namespace, svcName string) (*api.PersistentVolumeClaim, error) {
+	capacity, err := resource.ParseQuantity("10Gi")
+	if err != nil {
+		return nil, err
+	}
+
+	pvc := &api.PersistentVolumeClaim{
+		ObjectMeta: api.ObjectMeta{
+			Name:      fmt.Sprintf("%s-etcd-claim", svcName),
+			Namespace: namespace,
+			Labels: map[string]string{
+				"app": "federated-cluster",
+			},
+			Annotations: map[string]string{
+				"volume.alpha.kubernetes.io/storage-class": "yes",
+			},
+		},
+		Spec: api.PersistentVolumeClaimSpec{
+			AccessModes: []api.PersistentVolumeAccessMode{
+				api.ReadWriteOnce,
+			},
+			Resources: api.ResourceRequirements{
+				Requests: api.ResourceList{
+					api.ResourceStorage: capacity,
+				},
+			},
+		},
+	}
+
+	return clientset.Core().PersistentVolumeClaims(namespace).Create(pvc)
+}
+
+func createAPIServer(clientset *client.Clientset, namespace, name, credentialsName, pvcName string, ips []string) (*extensions.Deployment, error) {
+	command := []string{
+		"/hyperkube",
+		"federation-apiserver",
+		"--bind-address=0.0.0.0",
+		"--etcd-servers=http://localhost:2379",
+		"--service-cluster-ip-range=10.0.0.0/16",
+		"--secure-port=443",
+		"--token-auth-file=/etc/federation/apiserver/known_tokens.csv",
+		"--basic-auth-file=/etc/federation/apiserver/basic_auth.csv",
+		"--client-ca-file=/etc/federation/apiserver/ca.crt",
+		"--tls-cert-file=/etc/federation/apiserver/server.crt",
+		"--tls-private-key-file=/etc/federation/apiserver/server.key",
+	}
+	// Since only one IP address can be specified as advertise address, we arbitrarily pick the first availabe IP address
```

Comment:
In that case probably better to only pass one ip into this function, otherwise the caller will be surprised.